gnocchi/rest/__init__.py

```diff
@@ -1497,6 +1497,65 @@ def post(self):
 
         pecan.response.status = 202
 
+    @pecan.expose('json')
+    def get_all(self, **kwargs):
+        # Check RBAC policy
+        metric_ids = arg_to_list(kwargs.get('metric', []))
+        metrics = pecan.request.indexer.list_metrics(ids=metric_ids)
+        missing_metric_ids = (set(metric_ids)
+                              - set(six.text_type(m.id) for m in metrics))
+        if missing_metric_ids:
+            # Return one of the missing one in the error
+            abort(404, storage.MetricDoesNotExist(
+                missing_metric_ids.pop()))
+
+        for metric in metrics:
+            enforce("get metric", metric)
+
+        start = kwargs.get('start')
+        if start is not None:
+            try:
+                start = utils.to_datetime(start)
+            except Exception:
+                abort(400, "Invalid value for start")
+
+        stop = kwargs.get('stop')
+        if stop is not None:
+            try:
+                stop = utils.to_datetime(stop)
+            except Exception:
+                abort(400, "Invalid value for stop")
+
+        aggregation = kwargs.get('aggregation', 'mean')
+        if (aggregation
+           not in archive_policy.ArchivePolicy.VALID_AGGREGATION_METHODS):
+            abort(
+                400,
+                'Invalid aggregation value %s, must be one of %s'
+                % (aggregation,
+                   archive_policy.ArchivePolicy.VALID_AGGREGATION_METHODS))
+
+        granularity = kwargs.get('granularity')
+        if granularity is not None:
+            try:
+                granularity = Timespan(granularity)
+            except ValueError as e:
+                abort(400, e)
+
+        metric_batch = {}
+        try:
+            for metric in metrics:
+                measures = pecan.request.storage.get_measures(
+                    metric, start, stop, aggregation, granularity)
+                metric_batch[str(metric.id)] = [
+                    (timestamp.isoformat(), offset, v)
+                    for timestamp, offset, v in measures]
+        except (storage.GranularityDoesNotExist,
+                storage.AggregationDoesNotExist) as e:
+            abort(404, e)
+
+        return metric_batch
```

Comment:
I'm worried that this could create a huge list that uses a lot of memory.

IIRC there's a way to return an iterable here rather than a finished list. That'd allow the Web server to iterate on it as the client is reading.